sql/src/main/java/io/crate/jobs/AbstractExecutionSubContext.java

```diff
@@ -0,0 +1,151 @@
+ * permissions and limitations under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.jobs;
+
+import org.elasticsearch.common.logging.ESLogger;
+import org.elasticsearch.common.logging.Loggers;
+
+import javax.annotation.Nullable;
+import java.util.concurrent.CancellationException;
+
+public abstract class AbstractExecutionSubContext implements ExecutionSubContext, ExecutionState {
+
+    protected static final ESLogger LOGGER = Loggers.getLogger(AbstractExecutionSubContext.class);
+
+    protected final SubExecutionContextFuture future = new SubExecutionContextFuture();
+    private final int id;
+
+    protected JobExecutionContext context;
+
+    // TODO: remove this flag when ExecutionState is obsolete
+    private volatile boolean isKilled = false;
+
+    protected AbstractExecutionSubContext(int id) {
+        this.id = id;
+    }
+
+    @Override
+    public void jobExecutionContext(JobExecutionContext context) {
+        this.context = context;
+    }
+
+    public int id() {
+        return id;
+    }
+
+    protected void innerPrepare() {
+    }
+
+    @Override
+    final public void prepare() {
+        if (!future.closed()) {
+            LOGGER.trace("preparing {}: {}", this, id);
+            try {
+                innerPrepare();
+            } catch (Throwable t) {
+                close(t);
+            }
+        }
+    }
+
+    protected void innerStart() {
+    }
+
+    @Override
+    final public void start() {
+        if (!future.closed()) {
+            LOGGER.trace("starting {}: {}", this, id);
+            try {
+                innerStart();
+            } catch (Throwable t) {
+                close(t);
+            }
+        }
+    }
+
+    protected void innerClose(@Nullable Throwable t) {
+    }
+
+    protected boolean close(@Nullable Throwable t) {
+        if (future.firstClose()) {
+            LOGGER.trace("closing {}: {}", this, id);
+            try {
+                innerClose(t);
+            } catch (Throwable t2) {
+                future.close(t2);
+            } finally {
+                cleanup();
+            }
+            future.close(t);
+            return true;
+        }
+        return false;
+    }
+
+    @Override
+    final public void close() {
+        close(null);
+    }
+
+    protected void innerKill(@Nullable Throwable t) {
+    }
+
+    @Override
+    final public void kill(@Nullable Throwable t) {
+        if (future.firstClose()) {
+            if (t == null) {
+                t = new CancellationException();
+            }
+            try {
+                LOGGER.trace("killing {}: {}", this, id);
+                innerKill(t);
+            } catch (Throwable t2) {
+                future.close(t2);
```

Comment:
if this is the case future.close is called twice with different exceptions, is that okay?